tasks/sst_metrics_utils/filtermatches.py

```diff
@@ -0,0 +1,71 @@
+import numpy as np
+from lsst.afw.table import GroupView
+
+
+def filterMatches(matchedCatalog, snrMin=None, snrMax=None,
+                  extended=None, doFlags=None, isPrimary=None,
+                  psfStars=None, photoCalibStars=None,
+                  astromCalibStars=None):
+
+    if snrMin is None:
+        snrMin = 50.0
+    if snrMax is None:
+        snrMax = np.Inf
+    if extended is None:
+        extended = False
+    if doFlags is None:
+        doFlags = True
+    nMatchesRequired = 2
+    if isPrimary is None:
+        isPrimary = True
+    if psfStars is None:
+        psfStars = False
+    if photoCalibStars is None:
+        photoCalibStars = False
+    if astromCalibStars is None:
+        astromCalibStars = False
+
+    matchedCat = GroupView.build(matchedCatalog)
+    magKey = matchedCat.schema.find('slot_PsfFlux_mag').key
+
+    def nMatchFilter(cat):
+        if len(cat) < nMatchesRequired:
+            return False
+        return np.isfinite(cat.get(magKey)).all()
+
+    def snrFilter(cat):
+        # Note that this also implicitly checks for psfSnr being non-nan.
+        snr = cat.get('base_PsfFlux_snr')
+        ok0, = np.where(np.isfinite(snr))
+        medianSnr = np.median(snr[ok0])
+        return snrMin <= medianSnr and medianSnr <= snrMax
+
+    def ptsrcFilter(cat):
+        ext = cat.get('base_ClassificationExtendedness_value')
+        # Keep only objects that are flagged as "not extended" in *ALL* visits,
+        # (base_ClassificationExtendedness_value = 1 for extended, 0 for point-like)
+        if extended:
+            return np.min(ext) > 0.9
+        else:
+            return np.min(ext) < 0.9
+
+    def flagFilter(cat):
+        if doFlags:
+            flag_sat = cat.get("base_PixelFlags_flag_saturated")
+            flag_cr = cat.get("base_PixelFlags_flag_cr")
+            flag_bad = cat.get("base_PixelFlags_flag_bad")
+            flag_edge = cat.get("base_PixelFlags_flag_edge")
+            return np.logical_not(np.any([flag_sat, flag_cr, flag_bad, flag_edge]))
+        else:
+            return True
+
+    def isPrimary(cat):
+        if isPrimary:
+            flag_primary = cat.get("detect_isPrimary")
```

Comment:
It looks like `flag_primary` will be an array rather than a single boolean

Reply:
That's the desired behavior for all of the filter functions - they should return an array satisfying the requested condition.

Reply:
Oops, I was mistaken on two fronts. 1. This indeed should return an array, so proper return statement should include np.all(flag_primary), because 2. I realized that your other comment was correct. The filters only work if *all* matches meet the criteria. I played with it and struggled to figure out how to make it return only the "good" matches without be so harsh. Future upgrade?

Reply:
Got it. I think this will eventually become important because we will have 100+ visits of the same field, and so we would become more susceptible to a single poor-quality measurement